src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/RequestHeaderValueResolver.php

```diff
@@ -0,0 +1,95 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpKernel\Controller\ArgumentResolver;
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Attribute\MapRequestHeader;
+use Symfony\Component\HttpKernel\Controller\ValueResolverInterface;
+use Symfony\Component\HttpKernel\ControllerMetadata\ArgumentMetadata;
+use Symfony\Component\HttpKernel\Exception\HttpException;
+use Symfony\Component\Serializer\Exception\PartialDenormalizationException;
+use Symfony\Component\Serializer\Normalizer\DenormalizerInterface;
+use Symfony\Component\Serializer\SerializerInterface;
+use Symfony\Component\Validator\ConstraintViolationList;
+use Symfony\Component\Validator\Exception\ValidationFailedException;
+use Symfony\Component\Validator\Validator\ValidatorInterface;
+
+class RequestHeaderValueResolver implements ValueResolverInterface
+{
+    /**
+     * @see DenormalizerInterface::COLLECT_DENORMALIZATION_ERRORS
+     */
+    private const CONTEXT_DESERIALIZE = [
+        'collect_denormalization_errors' => true,
+    ];
+
+    public function __construct(
+        private readonly SerializerInterface&DenormalizerInterface $serializer,
+        private readonly ?ValidatorInterface $validator = null,
+    ) {
+    }
+
+    public function resolve(Request $request, ArgumentMetadata $argument): iterable
+    {
+        if (!$attribute = $argument->getAttributesOfType(MapRequestHeader::class)[0] ?? null) {
+            return [];
+        }
+
+        $headers = [];
+        $requestHeaders = $request->headers->all();
+
+        array_walk($requestHeaders, function ($value, $key) use (&$headers) {
+            $headers[$key] = implode(',', $value);
+        });
+
+        $type = $argument->getType();
+
+        if ('string' === $type || 'array' === $type) {
+            $name = $attribute->name ?? $argument->getName();
```

Comment:
I had to read the code to get the relation between the type of the parameter and the name of the attribute.
It's not obvious to me that typehinting for an object means "name" is ignored. We should at least throw when it's set to tell it shouldn't be here.
But it could be preferable to split this attribute in two: MapRequestHeader and MapRequestHeaders?

Reply:
@nicolas-grekas I think that's a sound idea! @StevenRenaux Are you still interested in working on this?

Reply:
Yes still needed, because if you write something as

```php
    #[Route(path: '/', name: 'test')]
    public function test(
        #[MapRequestHeader] string $accept,
    ): Response
    {
        // rest of your code
    }
```

name will be formed like this

```php

            $attribute->name, // null
            $argument->getName() // "accept" depending of you var name
```

And if you put a name

```php
    #[Route(path: '/', name: 'test')]
    public function test(
        #[MapRequestHeader('user-agent')] string $accept,
    ): Response
    {
        // rest of your code
    }
```

```php

            $attribute->name, // "user-agent"
            $argument->getName() // "accept" depending of you var name
```

Reply:
I was thinking about the `if ()` just above. sorry that it wasn't clear. anyway, it doesn't apply anymore with your latest changes :ok_hand: